include/igl/smooth_corner_adjacency.cpp

```diff
@@ -0,0 +1,149 @@
+#include "smooth_corner_adjacency.h"
+#include "vertex_triangle_adjacency.h"
+#include "matlab_format.h"
+#include "parallel_for.h"
+#include "unzip_corners.h"
+#include <iostream>
+
+void igl::smooth_corner_adjacency(
+  const Eigen::MatrixXd & V,
+  const Eigen::MatrixXi & F,
+  const double corner_threshold_radians,
+  Eigen::VectorXi & CI,
+  Eigen::VectorXi & CC)
+{
+  typedef double Scalar;
+  typedef Eigen::Index Index;
+  Eigen::Matrix<Index,Eigen::Dynamic,1> VF,NI;
+  igl::vertex_triangle_adjacency(F,V.rows(),VF,NI);
+  // unit normals
+  Eigen::Matrix<Scalar,Eigen::Dynamic,3,Eigen::RowMajor> FN(F.rows(),3);
+  //for(Index f = 0;f<F.rows();f++)
+  igl::parallel_for(F.rows(),[&](const Index f)
+  {
+    const Eigen::Matrix<Scalar,1,3> v10 = V.row(F(f,1))-V.row(F(f,0));
+    const Eigen::Matrix<Scalar,1,3> v20 = V.row(F(f,2))-V.row(F(f,0));
+    const Eigen::Matrix<Scalar,1,3> n = v10.cross(v20);
+    const Scalar a = n.norm();
+    FN.row(f) = n/a;
+  },10000);
+
+  // number of faces
+  const Index m = F.rows();
+  // valence of faces
+  const Index n = F.cols();
+  assert(n == 3);
+
+  CI.resize(m*n*8);
+  CI.setConstant(-1);
+  Index ncc = 0;
+  Index ci = -1;
+  // assumes that ci is strictly increasing and we're appending to CI
+  const auto append_CI = [&](Index nf)
+  {
+    // make room
+    if(ncc >= CI.size()) { CI.conservativeResize(CI.size()*2+1); }
+    CI(ncc++) = nf;
+    CC(ci+1)++;
+  };
+  CC.resize(m*3+1);
+  CC.setConstant(-1);
+  CC(0) = 0;
+
+  const Scalar cos_thresh = cos(corner_threshold_radians);
+  // parallelizing this probably requires map-reduce
+  for(Index i = 0;i<m;i++)
+  {
+    // Normal of this face
+    const auto & fnhat = FN.row(i);
+    // loop over corners
+    for(Index j = 0;j<n;j++)
+    {
+      // increment ci
+      ci++;
+      assert(ci == i*n+j);
+      CC(ci+1) = CC(ci);
+      const auto & v = F(i,j);
+      for(int k = NI(v); k<NI(v+1); k++)
+      {
+        const int nf = VF(k);
+        const auto & ifn = FN.row(nf);
+        // dot product between face's normal and other face's normal
+        const Scalar dp = fnhat.dot(ifn);
+        // if difference in normal is slight then add to average
+        if(dp > cos_thresh)
+        {
+          append_CI(nf);
+        }
+      }
+    }
+  }
+  CI.conservativeResize(ncc);
+}
+
+
+void igl::smooth_corner_adjacency(
```

Comment:
Why not call the other function with a threshold of 0 ? (and using a `>=` comparison).

Reply:
I don't understand this comment.

Reply:
Well you've basically implemented the function `smooth_corner_adjacency` twice. But the variant without the angle threshold can simply call the first function with `corner_threshold_radians == -M_PI`, provided you change the condition `if(dp > cos_thresh)` to a `if(dp >= cos_thresh * fnhat.norm() * ifn.norm())`. Well due to floating point error that might not be entirely correct, but you could use a lambda to implement the generic function instead.

Reply:
Sorry for the confusion, Jeremie. These two functions are not related that way.

I'll fix the header to document the second overload. It is purely combinatorial based on indexed normals (e.g., coming from a .obj).

Reply:
Oh ok, that makes sense.